src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ProxyTypeMapNode.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal sealed class ProxyTypeMapNode(TypeDesc typeMapGroup, IEnumerable<KeyValuePair<TypeDesc, TypeDesc>> mapEntries) : DependencyNodeCore<NodeFactory>, IProxyTypeMapNode
+    {
+        public TypeDesc TypeMapGroup { get; } = typeMapGroup;
+
+        public IEnumerable<KeyValuePair<TypeDesc, TypeDesc>> MapEntries => mapEntries;
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+
+        public override bool HasDynamicDependencies => false;
+
+        public override bool HasConditionalStaticDependencies => true;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public int ClassCode => 779513676;
+
+        public int CompareToImpl(ISortableNode other, CompilerComparer comparer) => comparer.Compare(TypeMapGroup, ((ProxyTypeMapNode)other).TypeMapGroup);
+
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory context)
+        {
+            List<CombinedDependencyListEntry> entries = [];
+            foreach (var (key, value) in mapEntries)
+            {
+                entries.Add(new CombinedDependencyListEntry(
+                    context.MaximallyConstructableType(value),
```

Comment:
Just double checking: in https://github.com/dotnet/runtime/issues/115323, we said that the proxy type map can also be used to effectively implement a fast type -> type lookup for any type of key type, including interface types (see https://github.com/dotnet/runtime/issues/115323#issuecomment-2889781663). In particular, we'll be using this in CsWinRT (with a dedicated type map group) specifically for the lookup of (potentially generic constructed types) interface types to IDIC implementation types. Will this already also correctly handle source types being interface types? Should we add unit tests for this specific scenario too, if not already there?

Also feel free to mark #115323 as also being resolved with this PR, assuming this scenario will already work 🙂

Reply:
Yep this will work for interface types. Because of IDIC, we need to have a maximally constructable node for interface types:

https://github.com/dotnet/runtime/blob/c160acf0f2d773c6a30271c76e0addda5dcb508b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs#L311-L315

As a result, even if we were able to optimize that case above for types that can't ever return a valid result for IDIC, it would be an effective lookup solution for IDIC implementation types,

Reply:
Reading through that thread again, there's some interesting test cases that Michal shared. I'll add some testing for those scenarios.

Reply:
Oh I see. Mmh question: can we restrict that to casts where the target object might actually be `IDynamicInterfaceCastable`? I.e. if the target object is a sealed type or a value type (or a constrained generic that's instantiated over a value type) that does not implement IDIC, presumably we could skip this and root fewer things, right?

Reply:
Added tests for the expected interface experiences.